README.md
# Pantry

[![Build Status](https://travis-ci.org/ocean/pantry.svg)](https://travis-ci.org/ocean/pantry)

A JSON/XML resource caching library based on Request

## Introduction

Pantry is an HTTP client cache used to minimize requests for external JSON and XML feeds.  Pantry will cache resources to minimize round trips, utilizing the local cache when available and refresh the cache (asynchronously) as needed.  (As of 0.7.x, Pantry can also proxy cache any raw resource.) 

As with any of our projects, constructive criticism is encouraged.

## Installing

Just grab [node.js](http://nodejs.org/#download) and [npm](http://github.com/isaacs/npm) and you're set:

	npm install pantry
	
Pantry uses the amazing [Request](https://github.com/mikeal/request) library for all HTTP requests and [xml2js](https://github.com/Leonidas-from-XIV/node-xml2js) to parse xml resources.

## Using

Work in progress.  See the examples for the time being.

To utilize the pantry, simply require it, optionally override some default values, and then request your resource(s) via the fetch method

	var pantry = require('pantry');
	pantry.configure({ shelfLife: 5 });
	
	pantry.fetch({ uri: 'http://search.twitter.com/search.json?q=winning'}, function (error, item, contentType) {
		console.log(item.results[0].text);
	});

At this item, the following configuration options can be specified:

* shelfLife - number of seconds before a resource reaches it's best-before-date
* maxLife - number of seconds before a resource spoils
* caseSensitive - URI should be considered case sensitive when inferring cache key
* verbosity - possible values are 'silly', 'debug, 'verbose', info', 'warn' and 'error'  (default is 'error' for production systems)
* parser - possible values are 'json', 'xml', or 'raw'  (default is undefined, in which auto-detection by content-type header is attempted)
* ignoreCacheControl - do not utilize the cache-control header to override the cache configuration if present (default is false)
* cacheBuster - adds the provided query string parameter name with a cachebusting timestamp to the request
* xmlOptions - options passed into the [xml2js](https://github.com/Leonidas-from-XIV/node-xml2js) parser (default is  {explicitRoot: false})

When you request a resource from the pantry, a couple interesting things happen.  If the item is available in the pantry, and hasn't 'spoiled', it will be returned immediately via the callback.  If it has expired (it's beyond its best before date) but hasn't spoiled, it will still be returned and then refreshed in the background.

> Ode to my immigrant mother:  The best before date is treated as a recommendation.  If it hasn't visibly spoiled it's probably still good to use, so use it until we have a chance to go shopping.  Especially if it's salad dressing, that stuff never goes bad as long as it's in the fridge.

If the resource isn't available in the pantry, or the item has spoiled, then the item will be retrieved immediately and won't be passed on to the callback method until we have the resource on hand.

Pantry will also ensure that we don't fetch the same resource multiple times in parallel.  If a resource is already being requested, additional requests for that same resource will hook into the same completion event for the original request.

If you wish to manually remove an item from the pantry, you can do this by calling the `remove()` method with the URI of the resource you wish to remove:

    pantry.remove('http://search.twitter.com/search.json?q=winning');

## Storage

The latest version of Pantry ( >= 0.3 ) supports the ability to plug the caching storage engine of your choice.  By default, pantry will utilize the MemoryStorage plugin, which will (of all things) cache items locally in memory.  

To specify an alternate storage engine, or to provide custom configuration for the default memory storage, simply assign the new storage engine to pantry via the .storage property.

### MemoryStorage

The constructor for MemoryStorage takes two parameters:

* config - hash of configuration properties (see below)
* verbosity - controls the level of logging (default is 'info')

The following configuration properties are allowed for MemoryStorage

* capacity - the maximum number of items to keep in the pantry (default is 1000)
* ideal - when cleaning up, the ideal number of items to keep in the pantry (default is 90% of capacity)


Example:

	var pantry = require('pantry');
		, MemoryStorage = require('pantry/lib/pantry-memory');

	pantry.storage = new MemoryStorage({
	  capacity: 18,
	  ideal: 12
	}, 'debug');

Note that the ideal must be set to a value which is between 10% and 90% of capacity.  Every time an item is added to pantry, we ensure we haven't reached capacity.  If we have, then we first start with throwing out any spoiled items.  After that, if we are still above capacity we will get rid of the expired items, and if we're really desperate we will need to throw out some good items just to make room.

### RedisStorage

A simply plugin for Redis is included with Pantry.  Note that since use of Redis is optional, the required client (redis) is not included in the package dependencies.  You must include it in your own application's dependencies and/or manually install it via npm install redis

The constructor for RedisStorage takes four parameters:

* port - the redis server port (default is 6379)
* host - the redis server host name (default is 'localhost')
* options - hash of configuration properties (see below)
* verbosity - controls the level of logging (default is 'info')

The following configuration properties are allowed for RedisStorage

* auth - the password / authentication key for the redis server (default is null)

Example:

	var pantry = require('pantry')
		, RedisStorage = require('pantry/lib/pantry-redis');

	pantry.storage = new RedisStorage(6379, 'localhost', null, 'debug');

### MemcachedStorage

A simply plugin for Memcached is also included with Pantry.  Note that since use of Memcached is optional, the required client (memcached) is not included in the package dependencies.  You must include it in your own application's dependencies and/or manually install it via npm install memcached

The constructor for MemcachedStorage takes three parameters:

* servers - a string or array of strings identifying the Memcached server(s) to use
* options - hash of memcached configuration properties (see [here](https://github.com/3rd-Eden/node-memcached#readme) for more details)
* verbosity - controls the level of logging (default is 'info')

Example:

	var pantry = require('../src/pantry')
		, MemcachedStorage = require('../src/pantry-memcached');

	pantry.storage = new MemcachedStorage('localhost:11211', {}, 'debug');
	

## SOAP Support

Support for SOAP requests has been dropped from this version due to the many security issues with dependencies, and the fact that I don't need SOAP support.
	
## Upgrading

As of v0.4.x, we now use v0.2.x of the xml2js library.  This has significantly changed the default parsing options.  You can easily revert to the xml2js v0.1 parsing options as described [here](https://github.com/Leonidas-from-XIV/node-xml2js)

## Roadmap

* Better handling of not-GET requests
* Ability to execute array of requests in parallel
* Support for cookies (including cache key)

## Created and managed by

* Edward de Groot
* Keith Benedict


History.md
1.0.0 / 2016-12-12
==================
  * Dropped support for SOAP.

0.7.8 / 2016-12-09
==================
  * Updated request and soap libraries.
  * Fixed some broken tests.

0.7.7 / 2016-11-28
==================
  * Cumulative updates of several dependencies: soap, mocha, winston, hawk.
  * Added Travis CI integration for testing.
  * Fixed Object.hasOwnProperty() bug in lib/pantry.js.

0.7.6 / 2016-01-07
==================
  * Added Pantry.remove() function to remove individual items from the Pantry cache.
  * Added usage description of the function to the README.
  * Added Mocha/Should tests for the remove() function, using a new testing JSON URL.
  * Corrected some typos in the test file and extended some test timeouts as they were failing.
  * Admin changes: bumped version number, added history changes and updated contributor list.

0.7.5 / 2015-02-20
==================
  * Fix for parsing maxLife and shellLife as integers always.
  * Updated dependencies for support of Node > 8.0
  * Soap@0.2.7 retained due to changes in the return results that break code dependant on this version or earlier
  * Updated license date

0.7.4 / 2014-06-16
==================
  * Fix for Date comparison in Expired and Spoiled checks when using third party memory storage.

0.7.3 / 2014-05-06
===================
  * Bumped request dependecy version

0.7.2 / 2014-01-06
===================
  * Added cacheBuster config option

0.7.1 / 2014-01-06
===================
  * Re-applied missing ignoreCacheControl setting

0.7.0 / 2014-01-04
===================
  * Experimental support for caching any raw resource

0.5.3 / 2013-08-27
===================
  * Fixed variable undefined error causing application crashes in the 'done' event function

0.5.2 / 2013-07-17
===================
  * Added support for complete  data types in SOAP requests

0.5.1 / 2013-07-16
===================
  * Experimental support for SOAP requests

0.5.0 / 2013-07-16
===================
  * Converted from CoffeeScript to JavaScript
  * Now uses winston instead of coloured-log for logging

0.4.1 / 2012-10-11
===================
  * Update default logging with optional env setting

0.4.0 / 2012-10-05
===================
  * Upgraded dependencies
  * Requires Node v0.8.x
  * Default XML parsing has changed with xml2js v0.2

0.3.4 / 2012-10-04
===================

  * Fixing a bug in dependencies, xml2js version set to 0.1.14 as newer versions use incompatible versions of coffee-script.

0.3.3 / 2012-04-24
==================

  * Handle request library returning an object (no parsing needed)

0.3.2 / 2012-03-29
==================

  * Better handling of offline storage

0.3.1 / 2012-03-28
==================

  * Use MemoryStorage as backup if primary storage is unavailable

0.3.0 / 2012-03-21
==================

  * Unstable experimental release
  * Code has been completely refactored
  * Supports plugins for cache storage (defaults to MemoryStorage)
  * Added RedisStorage plugin
  * Added MemcachedStorage plugin
  * Switched to mocha testing framework

0.2.0 / 2011-10-03
==================

  * Stable production release
  * Reversed sort order within History.md

0.2.0beta2 / 2011-09-15
=======================

  * xml2js parsing options can now be passed in via xmlOptions
  * parsed xml no longer provides explicit arrays for child elements by default (but you can enable via xmlOptions)

0.2.0beta / 2011-09-14
======================

  * Now respects cache-control header
  * non-GET requests are not cached
  * parsed xml now uses explicit arrays for child elements
  * ability to utilize loose parsing when dealing with inexplicit or incorrect content-types

0.1.2 / 2011-09-06
===================

  * Bug Fix - Removed config = config
  * Convert from styout to coloured-log

0.1.1 / 2011-08-18
===================

  * Slight mod to node version requirements

0.1.0 / 2011-08-09
===================

  * Added return value to configuration method containing the actual config values.
  * Added getStock method that builds a data structure containing the stock count and current stock items
  * implemented test cases using kitkat

0.0.3 / 2011-08-03
==================

  * Normalizes URI to create consistent cache keys
  * Utilizes for styout for console messages (debug, info, errors etc.)

0.0.2 / 2011-07-28
==================

  * Major refactoring
  * Supports XML resources
  * Conditional GETs (ETag and/or Last-Modified)
  * Efficient handling of parallel requests for the same resource
  * Some error handling

0.0.1 / 2011-07-18
==================

  * Initial release


test/pantry-test.js

process.env.NODE_ENV = 'test';

var should = require('should')
  , pantry = require('../lib/pantry')
  , jsonURL = 'http://api.flickr.com/services/rest/?api_key=999a559d39e8a2c23397f740d53e4447&format=json&method=flickr.test.echo&nojsoncallback=1'
  , xmlURL = 'http://api.flickr.com/services/rest/?method=flickr.test.echo&api_key=999a559d39e8a2c23397f740d53e4447&format=rest'
  , anotherJsonURL = 'http://date.jsontest.com';


describe('pantry', function() {
  describe('configure', function() {
    
    it('should have default configuration', function() {
      var config;
      config = pantry.configure({});
      config.should.have.property('shelfLife', 60);
      config.should.have.property('maxLife', 300);
      config.should.have.property('caseSensitive', true);
    });
    
    it('should allow configuration overides', function() {
      var config;
      config = pantry.configure({
        caseSensitive: false
      });
      config.should.have.property('shelfLife', 60);
      config.should.have.property('maxLife', 300);
      config.should.have.property('caseSensitive', false);
    });
    
  });
  
  describe('generateKey', function() {
    
    it('should leave case alone if caseSensitive', function() {
      var key = pantry.generateKey({
        uri: 'http://search.twitter.com/Search.json?Q=sugar',
        caseSensitive: true
      });
      key.should.equal('http://search.twitter.com/Search.json?Q=sugar');
    });
    
    it('should lower case alone if not caseSensitive', function() {
      var key = pantry.generateKey({
        uri: 'http://search.twitter.com/Search.json?Q=sugar',
        caseSensitive: false
      });
      key.should.equal('http://search.twitter.com/search.json?q=sugar');
    });
    
    it('should rearrange parameters alphabetically', function() {
      var key = pantry.generateKey({
        uri: 'http://search.twitter.com/search.json?since=1234&q=sugar',
        caseSensitive: true
      });
      key.should.equal('http://search.twitter.com/search.json?q=sugar&since=1234');
    });
    
  });
  
  describe('hasSpoiled', function() {
    
    it('should identify empty resources as spoiled', function() {
      var resource = {};
      pantry.hasSpoiled(resource).should.be.true;
    });
    
    it('should identify old resources as spoiled', function() {
      var resource = {
        results: 'test',
        spoilsOn: new Date()
      };
      resource.spoilsOn.setHours(resource.spoilsOn.getHours() - 1);
      pantry.hasSpoiled(resource).should.be.true;
    });
    
    it('should identify stale resources as good', function() {
      var resource = {
        results: 'test',
        spoilsOn: new Date()
      };
      resource.spoilsOn.setHours(resource.spoilsOn.getHours() + 1);
      pantry.hasSpoiled(resource).should.be.false;
    });
    
  });
  
  describe('hasExpired', function() {
    
    it('should identify empty resources as expired', function() {
      var resource = {};
      pantry.hasExpired(resource).should.be.true;
    });
    
    it('should identify spoiled resources as expired', function() {
      var resource = {
        results: 'test',
        spoilsOn: new Date()
      };
      
      resource.spoilsOn.setHours(resource.spoilsOn.getHours() - 1);
      pantry.hasExpired(resource).should.be.true;
    });
    
    it('should identify expired but not spoiled resources as expired', function() {
      var resource = {
        results: 'test',
        spoilsOn: new Date(),
        bestBefore: new Date()
      };
      resource.spoilsOn.setHours(resource.spoilsOn.getHours() + 1);
      resource.bestBefore.setHours(resource.bestBefore.getHours() - 1);
      pantry.hasExpired(resource).should.be.true;
    });
    
    it('should identify fresh resources as good', function() {
      var resource = {
        results: 'test',
        spoilsOn: new Date(),
        bestBefore: new Date()
      };
      resource.spoilsOn.setHours(resource.spoilsOn.getHours() + 2);
      resource.bestBefore.setHours(resource.bestBefore.getHours() + 1);
      pantry.hasExpired(resource).should.be.false;
    });
    
  });
  
  describe('fetch', function() {
    it('should return a JSON resource as an object', function(done) {
      this.timeout(2000);
      pantry.fetch(jsonURL, function(error, results) {
        results.should.be.a.Object();
        done(error);
      });
    });
    
    it('should return an XML resource as an object', function(done) {
      this.timeout(2000);
      pantry.fetch(xmlURL, function(error, results) {
        results.should.be.a.Object();
        done(error);
      });
    });
    
    it('should return an error for non JSON/XML resources', function(done) {
      pantry.fetch('http://google.com', function(error, results) {
        should.exist(error);
        done();
      });
    });
    
    it('should return an error for non existent resources', function(done) {
      pantry.fetch('http://search.twitter.com/bad', function(error, results) {
        should.exist(error);
        done();
      });
    });
    
    it('should return an error for non existent server', function(done) {
      pantry.fetch('http://bad.twitter.com/search.atom?q=sugar', function(error, results) {
        should.exist(error);
        done();
      });
    });
    
    it('should return an error for malformed uri', function(done) {
      return pantry.fetch('bad://search.twitter.com/search.atom?q=sugar', function(error, results) {
        should.exist(error);
        done();
      });
    });
    
  });
  
  describe('remove', function() {
    before(function(done) {
      this.timeout(3000);
      return pantry.fetch(anotherJsonURL, function(error, results) {
        done(error);
      });
    });

    it('should remove a currently cached item if it exists', function(done) {
      var currentCacheCount = pantry.storage.stockCount;
      pantry.remove(anotherJsonURL);
      pantry.storage.stockCount.should.equal(currentCacheCount - 1);
      pantry.storage.get(anotherJsonURL, function(error, resource) {
        should.not.exist(error);
        should.not.exist(resource);
      });
      done();
    });

    it('should return null if the cached item doesn\'t exist', function(done) {
      should.not.exist(pantry.remove(anotherJsonURL));
      done();
    });

  });

  return describe('storage', function() {
    it('should cache a previously requested resource', function(done) {
      pantry.storage.get(jsonURL, function(error, resource) {
        should.exist(resource);
        resource.should.have.property('options');
        resource.should.have.property('results');
        resource.should.have.property('firstPurchased').with.instanceof(Date);
        resource.should.have.property('lastPurchased').with.instanceof(Date);
        resource.should.have.property('bestBefore').with.instanceof(Date);
        resource.should.have.property('spoilsOn').with.instanceof(Date);
        resource.options.should.have.property('key');
        resource.options.should.have.property('uri');
        resource.options.should.have.property('shelfLife');
        resource.options.should.have.property('maxLife');
        done(error);
      });
    });
    
    it('should return cached results for subsequent calls', function(done) {
      pantry.storage.get(jsonURL, function(first_error, resource) {
        should.exist(resource);
        pantry.fetch(jsonURL, function(second_error, second_results) {
          second_results.should.eql(resource.results);
          done(first_error || second_error);
        });
      });
    });
    
  });
});

lib/pantry.js
var url = require('url')
  , querystring = require('querystring')
  , request = require('request')
  , xml2js = require('xml2js')
  , EventEmitter = require('events').EventEmitter
  , MemoryStorage = require('./pantry-memory')
  , winston = require('winston')
  , inProgress = {};


function getLogLevel() {
  switch (process.env.NODE_ENV) {
    case 'production':
      return 'info';
    case 'test':
      return 'error';
    default:
      return 'silly';
  }
}

config = {
  shelfLife: 60,
  maxLife: 300,
  caseSensitive: true,
  ignoreCacheControl: false,
  cacheBuster: null,
  verbosity: process.env.PANTRY_VERBOSITY || getLogLevel(),
  xmlOptions: {
    explicitRoot: false
  }
};

var log = new (winston.Logger)({
  transports: [
    new (winston.transports.Console)({
      colorize: true,
      timestamp: true,
      level: config.verbosity
    })
  ]
});

var Pantry = module.exports = {
  storage: null,
  backup: new MemoryStorage(config, config.verbosity)
};

Pantry.configure = function(options) {
  for (var k in options) {
    config[k] = options[k];
  }
  //log = new Log(config.verbosity);
  return config;
};

Pantry.fetch = function(options, callback) { 
  if (typeof options === 'string') {
    options = {
      uri: options
    };
  }
  
  for (var k in config) {
    if (options[k] == null)
      options[k] = config[k];
  }
  
  if (options.method != null && options.method !== 'GET') {
    options.maxLife = 0;
  }
  
  if (options.key == null)
    options.key = Pantry.generateKey(options);

  // These two lines solve the problem that a user may have submitted the values as strings, which affects cache lifetime
  options.maxLife = parseInt(options.maxLife);
  options.shelfLife = parseInt(options.shelfLife);

  Pantry.fromCache(options.key, function(error, resource) {
    resource = resource || {};
    resource.options = options;
    
    if (!Pantry.hasSpoiled(resource)) {
      log.info("using cached data: " + resource.options.key);
      callback(null, resource.results, resource.contentType);
      callback = null;
    }
    
    if (Pantry.hasExpired(resource)) {
      var stock = inProgress[options.key];
      if (stock != null) {
        log.info("waiting for new data: " + resource.options.key);
        if (callback)
          return stock.once('done', callback);
      } else {
        log.info("requesting new data: " + resource.options.key);
        stock = new EventEmitter();
        inProgress[options.key] = stock;
        stock.resource = resource;
        
        if (callback)
          stock.once('done', callback);
          
        resource.options.headers = resource.options.headers || {};
        
        if (!resource.options.ignoreCacheControl && resource.etag != null)
          resource.options.headers['if-none-match'] = resource.etag;
          
        if (!resource.options.ignoreCacheControl && resource.lastModified != null) 
          resource.options.headers['if-modified-since'] = resource.lastModified;
        
        try {
          
          var uri = url.parse(options.uri, true);
          // apply cache buster if configured
          if (options.cacheBuster) {
            var ts = new Date();
            uri.query[options.cacheBuster] = ts.getTime();
            delete uri.search;
            options.uri = url.format(uri);
          }
          
          request(options, function(error, response, body) {
          
            if (error) {
              return Pantry.done(error, stock);
            } else {
              switch (response.statusCode) {
                case 304:
                  log.info("cached data still good: " + resource.options.key);
                  Pantry.done(null, stock);
                  break;
                
                case 200:
                  log.info("new data available: " + options.key);
                  resource.contentType = response.headers["content-type"];
                
                  if (response.headers['etag'] != null)
                    resource.etag = response.headers['etag'];
                
                  if (response.headers['last-modified'] != null)
                    resource.lastModified = response.headers['last-modified'];
                
                  if (options.parser === 'xml' || resource.contentType.search(/[\/\+]xml/) > 0) {
                    var start = body.indexOf('<');
                    if (start)
                      body = body.slice(start, body.length);
                  
                    var parser = new xml2js.Parser(options.xmlOptions);
                    parser.on('end', function(results) {
                      resource.results = results;
                      Pantry.done(null, stock);
                    });
                    parser.parseString(body);
                  
                  } else if (typeof body === 'string' && options.parser !== 'raw') {
                    try {
                      resource.results = JSON.parse(body);
                      Pantry.done(null, stock);
                    } catch (err) {
                      Pantry.done(err, stock);
                    }
                  } else {
                    resource.results = body;
                    Pantry.done(null, stock);
                  }
                  break;
                
                default:
                  return Pantry.done("Invalid Response Code (" + response.statusCode + ")", stock);
              }
            }
          });
        } catch (err) {
          return Pantry.done(err, stock);
        }
      }
    }
  });
};

Pantry.remove = function(options) {
  if (typeof options === 'string') {
    options = {
      uri: options
    };
  }

  if (options.key == null) {
    options.key = Pantry.generateKey(options);
  }

  if (Pantry.storage == null) {
    Pantry.storage = Pantry.backup;
  }

  return Pantry.storage.remove(options.key);
};

Pantry.fromCache = function(key, callback) {
  if (Pantry.storage == null)
    Pantry.storage = Pantry.backup;
  
  Pantry.storage.get(key, function(error, resource) {
    if (error) {
      log.error("Problems with primary storage " + key);
      log.error(error);
      Pantry.backup.get(key, function(error, resource) {
        callback(error, resource);
      });
    } else {
      callback(error, resource);
    }
  });
};

Pantry.done = function(err, stock) {
  delete inProgress[stock.resource.options.key];
  
  if (err) {
    log.error("" + err);
    stock.emit('done', err);
  } else {
    var resource = stock.resource;
    resource.lastPurchased = new Date();
    resource.firstPurchased = resource.firstPurchased || resource.lastPurchased;
    resource.bestBefore = new Date(resource.lastPurchased);
    resource.bestBefore.setSeconds(resource.bestBefore.getSeconds() + resource.options.shelfLife);
    resource.spoilsOn = new Date(resource.lastPurchased);
    resource.spoilsOn.setSeconds(resource.spoilsOn.getSeconds() + resource.options.maxLife);
    
    if (resource.options.maxLife === 0) {
      stock.emit('done', null, resource.results, resource.contentType);
    } else {
      stock.emit('done', null, resource.results, resource.contentType);
      Pantry.storage.put(resource, function(error) {
        if (error) {
          log.error("Could not cache resource " + resource.options.key + ": " + error);
          Pantry.backup.put(resource);
        }
      });
    }
  }
};

Pantry.generateKey = function(options) {
  var uri = url.parse(options.uri, true);
  var keys = [];
  
  for (var k in uri.query) {
    keys.push(k);
  }
  keys.sort();
  
  var query = {};
  keys.forEach(function(k) {
    if (Object.prototype.hasOwnProperty.call(uri.query, k))
      query[options.caseSensitive ? k : k.toLowerCase()] = uri.query[k];    
  });
  
  uri.search = querystring.stringify(query);
  if (!options.caseSensitive) {
    uri.pathname = uri.pathname.toLowerCase();
  }
  
  return url.format(uri);
};

Pantry.hasSpoiled = function(resource) {
  return !(resource.results != null) || (new Date()) > new Date(resource.spoilsOn);
};

Pantry.hasExpired = function(resource) {
  return Pantry.hasSpoiled(resource) || (new Date()) > new Date(resource.bestBefore);
};
